src/main/resources/testchipip/csrc/testchip_fesvr.h

```diff
@@ -0,0 +1,31 @@
+#ifndef __TESTCHIP_FESVR_H
+#define __TESTCHIP_FESVR_H
+
+#include <fesvr/tsi.h>
+#include <fesvr/htif.h>
+
+class testchip_fesvr_t : public tsi_t
+{
+ public:
+  testchip_fesvr_t(int argc, char** argv, bool has_loadmem);
+  virtual ~testchip_fesvr_t() {};
+
+  void write_chunk(addr_t taddr, size_t nbytes, const void* src) override;
+  void read_chunk(addr_t taddr, size_t nbytes, void* dst) override;
+  void load_program() {
+    switch_to_target();
+    is_loadmem = has_loadmem;
+    tsi_t::load_program();
+    is_loadmem = false;
+  }
+
+ protected:
+  virtual void load_mem_write(addr_t taddr, size_t nbytes, const void* src) { };
```

Comment:
There seems to be two ways loadmem can be implemented.

- The first approach is what firesim-fesvr does. `load_payload` calls `write_chunk`, which calls a separate `loadmem_write` method when in loadmem mode. 
- The second approach is what the current SimDRAM simulations do. The `SimDRAM` `mm_t` parses the args separately from htif, and when the `+loadmem` arg is seen, it initializes the simulated memory with the payload. In this implementation of loadmem, htif doesn't participate in the loadmem channel, so the `load_mem_write` method should be a no-op, which is the default behavior of these virtual functions.

The way SimDRAM loads memory seems to not adhere to the right way to do this, which is how firesim-fesvr loads mem. A more correct implementation would override the `loadmem_write` methods instead. This seems like it should be in a future PR.

Reply:
I think there is an issue with the `load_mem_write` implementation but I can't be sure since I haven't read an implementation yet.
The upstream support for this only changes how writing is done during elf loading/initialization, where as this change seems like it would affect all writes.

Reply:
The `write_chunk` and `read_chunk` overridden methods look at the `is_loadmem` flag to ensure that `loadmem_write/read` are only called during the loadmem phase. Outside of the loadmem phase, the inherited `tsi_t::read/write_chunk` methods are used.